### PR TITLE
[SPARK-38999][SQL] Refactor `FileSourceScanExec`: file scan physical node  

### PR DESCRIPTION
**Pull request by utkarsh39**
### What changes were proposed in this pull request?
The PR refactors `FileSourceScanExec` case class into a base trait `FileSourceScanLike` which is then subclassed by `FileSourceScanExec`. `FileSourceScanLike` contains basic functionality like metrics and file listing while the `FileSourceScanExec` contains execution specific code.

### Why are the changes needed?
Currently the code for `FileSourceScanExec` class, the physical node for the file scans is quite complex and lengthy making it slightly difficult to reason about.

### Does this PR introduce _any_ user-facing change?
No

### How was this patch tested?
Code refactor, existing tests should suffice.

### CODE REVIEW

**Comment by utkarsh39** [comment]
@cloud-fan @gengliangwang Can you please help review this?

File: sql/core/src/main/scala/org/apache/spark/sql/execution/DataSourceScanExec.scala [hunk 1]
```diff
@@ -509,16 +478,73 @@ case class FileSourceScanExec(
...
+    if (relation.fileFormat.isInstanceOf[ParquetSource]) {
+      conf.parquetVectorizedReaderEnabled
     } else {
-      Map.empty[String, SQLMetric]
+      false
     }
-  } ++ staticMetrics
+  }
+
+  lazy val inputRDD: RDD[InternalRow] = {
```
Comment (HyukjinKwon) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/DataSourceScanExec.scala:524]: I like this idea of separating planning/metrics stuff vs RDD/execution.

File: sql/core/src/main/scala/org/apache/spark/sql/execution/DataSourceScanExec.scala [hunk 2]
```diff
@@ -179,48 +177,34 @@ case class RowDataSourceScanExec(
...
+  def dataFilters: Seq[Expression]
+  // Disable bucketed scan based on physical query plan, see rule
+  // [[DisableUnnecessaryBucketedScan]] for details.
+  def disableBucketedScan: Boolean
+  // Bucket ids for bucket pruning.
+  def optionalBucketSet: Option[BitSet]
+  // Number of coalesced buckets.
+  def optionalNumCoalescedBuckets: Option[Int]
+  // Output attributes of the scan, including data attributes and partition attributes.
+  def output: Seq[Attribute]
```
Comment (cloud-fan) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/DataSourceScanExec.scala:194]: `DataSourceScanExec` extends `LeafExecNode`, and `def output: Seq[Attribute]` is already declared there.

File: sql/core/src/main/scala/org/apache/spark/sql/execution/DataSourceScanExec.scala [hunk 3]
```diff
@@ -179,48 +177,34 @@ case class RowDataSourceScanExec(
...
+  // Predicates to use for partition pruning.
+  def partitionFilters: Seq[Expression]
```
Comment (cloud-fan) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/DataSourceScanExec.scala:196]: nit: can we put it near `dataFilters`?

**Review by cloud-fan — APPROVED** [review]

**Comment by cloud-fan** [comment]
thanks, merging to master!

File: sql/core/src/main/scala/org/apache/spark/sql/execution/DataSourceScanExec.scala [hunk 4]
```diff
@@ -230,17 +214,28 @@ case class FileSourceScanExec(
...
+      Map.empty[String, SQLMetric]
+    }
+  } ++ staticMetrics
 
   /**
    * Send the driver-side metrics. Before calling this function, selectedPartitions has
    * been initialized. See SPARK-26327 for more details.
    */
-  private def sendDriverMetrics(): Unit = {
-    driverMetrics.foreach(e => metrics(e._1).add(e._2))
```
Comment (LuciferYang) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/DataSourceScanExec.scala:240]: why remove this line?

Comment (cloud-fan) [reply sql/core/src/main/scala/org/apache/spark/sql/execution/DataSourceScanExec.scala:240]: previously `driverMetrics` is not `Map[String, SQLMetric]`, but just `Map[String, Long]`, so we need to update the real `SQLMetric` here . Now the refactor makes `driverMetrics` real `SQLMetric`

Comment (LuciferYang) [reply sql/core/src/main/scala/org/apache/spark/sql/execution/DataSourceScanExec.scala:240]: thanks @cloud-fan

**Pull request closed without merging**